hta/common/trace_filter.py

```diff
@@ -0,0 +1,363 @@
+        name_column: Optional[str] = None,
+    ) -> None:
+        self.name_pattern: str = name_pattern
+        self.symbol_table: Optional[TraceSymbolTable] = symbol_table
+        self.name_column: Optional[str] = name_column
+
+    def __call__(
+        self, df: pd.DataFrame, symbol_table: Optional[TraceSymbolTable] = None
+    ) -> pd.DataFrame:
+        if df.empty:
+            logger.warning("The DataFrame is empty.")
+            return df
+
+        if symbol_table is None and self.symbol_table is None:
+            return NameStringColumnFilter(self.name_pattern)(df)
+        else:
+            _symbol_table = (
+                symbol_table
+                if symbol_table
+                else self.symbol_table
+                if self.symbol_table
+                else TraceSymbolTable()
+            )
+            return NameIdColumnFilter(self.name_pattern)(df, _symbol_table)
+
+
+class GPUKernelFilter(Filter):
+    """
+    A trace event filter class that extracts GPU kernel events from a DataFrame.
+    """
+
+    def __call__(
+        self, df: pd.DataFrame, symbol_table: Optional[TraceSymbolTable] = None
+    ) -> pd.DataFrame:
+        if "stream" not in df.columns:
+            logger.warning("DataFrame does not contain a 'stream' column.")
+            return df
+
+        return df.loc[df["stream"] > 0]
+
+
+class CPUOperatorFilter(Filter):
+    """
+    A trace event filter class that extracts CPU operator events from a DataFrame.
+    """
+
+    def __call__(
+        self, df: pd.DataFrame, symbol_table: Optional[TraceSymbolTable] = None
+    ) -> pd.DataFrame:
+        if "stream" not in df.columns:
+            logger.warning("DataFrame does not contain a 'stream' column.")
+            return df
+
+        return df.loc[df["stream"] == -1]
+
+
+class CompositeFilter(Filter):
+    """
+    A composite trace filter class that applies a sequence of filters to a DataFrame.
+
+    This class allows combining multiple filter objects and applying them sequentially to a DataFrame.
+
+    Attributes:
+        filters (List[Filter]): A list of filter objects to be applied.
+    """
+
+    def __init__(self, filters: List[Filter]) -> None:
+        """
+        Initialize a CompositeFilter object with a list of filter objects.
+
+        Args:
+            filters (List[Filter]): A list of filter objects.
+        """
+        for obj in filters:
+            if not isinstance(obj, Filter):
+                raise TypeError(
+                    "All elements in 'filters' must be instances of Filter or its subclasses."
+                )
+            if obj is self:
+                raise ValueError("CompositeFilter cannot contain itself as a filter.")
+
+        self.filters: List[Filter] = filters
+
+    def __call__(
+        self, df: pd.DataFrame, symbol_table: Optional[TraceSymbolTable] = None
+    ) -> pd.DataFrame:
+        for f in self.filters:
+            df = f(df, symbol_table)
```

Comment:
this implementation is sub-optimal as it will do `len(self.filters)` passes over the data frame. Ideally, this should be done implemented as a single pass over the dataframe.